src/oge/load_data.py

```diff
@@ -116,21 +121,215 @@ def load_cems_data(year):
+        ["plant_id_eia", "generator_id"]
+    )["report_date"].transform("min")
+
+    # drop any data that was reported prior to the earliest year
+    # only keep data for years <= the year
+    # this avoids using potentially preliminary early-release data
+    complete_gens = complete_gens[
+        (complete_gens["report_date"].dt.year >= earliest_year)
+        & (complete_gens["report_date"].dt.year <= year)
+    ]
+
+    # for any retired gens, forward fill the most recently available unit_id_pudl to the
+    # most recent available year
+    complete_gens["unit_id_pudl"] = complete_gens.groupby(
+        ["plant_id_eia", "generator_id"]
+    )["unit_id_pudl"].ffill()
+
+    # only keep the most recent entry for each generator
+    complete_gens = complete_gens.sort_values(
+        by=["plant_id_eia", "generator_id", "report_date"], ascending=True
+    ).drop_duplicates(subset=["plant_id_eia", "generator_id"], keep="last")
+
+    # remove generators that are proposed but not yet under construction, or cancelled
+    status_codes_to_remove = ["CN", "IP", "P", "L", "T"]
+    complete_gens = complete_gens[
+        ~complete_gens["operational_status_code"].isin(status_codes_to_remove)
+    ]
+
+    # remove generators that retired prior to the earliest year
+    complete_gens = complete_gens[
+        ~(
+            (complete_gens["operational_status_code"] == "RE")
+            & (complete_gens["generator_retirement_date"].dt.year < earliest_year)
+        )
+    ]
+
+    # remove generators that have no operating or retirement date, and the last time they
+    # reported data was prior to the current year. This is often proposed plants that are
+    # assigned a new plant_id_eia once operational
+    complete_gens = complete_gens[
+        ~(
+            (complete_gens["generator_operating_date"].isna())
+            & (complete_gens["generator_retirement_date"].isna())
+            & (complete_gens["report_date"].dt.year < year)
+        )
+    ]
+
+    ####################
+    # merge into complete_gens and fill missing operating dates with the EIA-860 data
+    generator_data_from_eia860 = load_raw_eia860_generator_dates_and_unit_ids(year)
+    complete_gens = complete_gens.merge(
+        generator_data_from_eia860,
+        how="left",
+        on=["plant_id_eia", "generator_id"],
+        validate="1:1",
+    )
+    complete_gens["generator_operating_date"] = complete_gens[
+        "generator_operating_date"
+    ].fillna(complete_gens["operating_date_eia"])
+    complete_gens = complete_gens.drop(columns="operating_date_eia")
+
+    #######################
+    # update the unit_id_eia_numeric to be one higher than the highest existing unit_id_pudl
+    # if unit_id_eia_numeric is NA, the updated value should also still be na
+    complete_gens["unit_id_eia_numeric"] = complete_gens[
+        "unit_id_eia_numeric"
+    ] + complete_gens.groupby("plant_id_eia")["unit_id_pudl"].transform("max").fillna(0)
```

Comment:
I am lost here. I don't see `unit_id_eia_numeric` being set previously in this function. The `complete_gens` data frame is declared and assigned via the `load_pudl_table` function where we keep multiple columns but not `unit_id_eia_numeric`. Then, I don't see where `unit_id_eia_numeric` is introduced.

Reply:
This column is part of `generator_data_from_eia860`, which is loaded in `load_raw_eia860_generator_dates_and_unit_ids()` on line 228